### PR TITLE
Profile

### PR DESCRIPTION
**Pull request by TarasOrlovskyi**

### CODE REVIEW

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java [hunk 1]
```diff
@@ -35,6 +35,19 @@ public boolean add(String email, String password) {
         return isAdded;
     }
 
+    @Override
+    public boolean edit(String oldEmail, String newEmail, String newPassword){
+        boolean isEdit = false;
+        if (newEmail != null && newPassword != null){
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java]: Add oldEmail!=null.

File: src/main/resources/templates/editProfile.html [hunk 2]
```diff
@@ -93,34 +93,80 @@
...
+                </script>
                 <label class="user__label">
-                  <input class="user__input" type="password" name="confirmNewPassword"
+                  <input class="user__input" type="password" name="confirmNewPassword" id="confirmNewPassword"
                          placeholder="Confirm new password" required>
                 </label>
+                <script language='javascript'>
+                  function matchPassword() {
+                    if(document.form_edit.newPassword.value != document.form_edit.confirmNewPassword.value) {
+                      document.getElementById('user_message').innerHTML="Sorry, the passwords don't match!";
```
Comment (Greengrer) [inline src/main/resources/templates/editProfile.html:161]: *without the(the is for singular)
Sorry, passwords don't match!

File: src/main/resources/templates/editProfile.html [hunk 3]
```diff
@@ -93,34 +93,80 @@
...
+                      document.form_edit.confirmNewPassword.focus();
+                      return false;
+                    }
+                    document.form_edit.submit();
+                  }
+                </script>
                 <label class="user__label">
-                  <input class="user__input" type="text" name="discogsUserName" placeholder="Discogs User Name">
+                  <input class="user__input" type="text" name="discogsUserName" placeholder="Discogs Username">
```
Comment (Greengrer) [inline src/main/resources/templates/editProfile.html:169]: I appreciate it. :)

**Review by Greengrer — CHANGES_REQUESTED** [review]

File: src/main/java/com/vinylteam/vinyl/Starter.java [hunk 4]
```diff
@@ -65,25 +65,33 @@ public static void main(String[] args) throws Exception {
...
+        EditProfileServlet editProfileServlet = new EditProfileServlet(securityService, userService);
         HomeServlet homeServlet = new HomeServlet();
 
         Resource resource = JarFileResource.newClassPathResource(RESOURCE_PATH);
         ServletContextHandler servletContextHandler = new ServletContextHandler(ServletContextHandler.SESSIONS);
         servletContextHandler.setErrorHandler(new DefaultErrorHandler());
         servletContextHandler.setBaseResource(resource);
 
+        servletContextHandler.addFilter(new FilterHolder(securityFilter),
+                "/*",
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/Starter.java]: Не понятен этот перенос, зачем он?

File: src/main/java/com/vinylteam/vinyl/web/filter/SecurityFilter.java [hunk 5]
```diff
@@ -14,17 +16,32 @@
     public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
         HttpServletRequest httpServletRequest = (HttpServletRequest) request;
         HttpServletResponse httpServletResponse = (HttpServletResponse) response;
+        String signIn = "/signIn";
+        String recoveryPassword = "/recoveryPassword";
+        String signUp = "/signUp";
+        String profile = "/profile";
+        String editProfile = "/editProfile";
+        String signOut = "/signOut";
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/filter/SecurityFilter.java]: Фильтр нужно переделать. Писать много, созвонимся

File: src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                attributes.put("email", email);
+                PageGenerator.getInstance().process("editProfile", attributes, response.getWriter());
+            } else {
+                response.sendRedirect("/");
+            }
+
+        } else {
+            response.sendRedirect("/");
+        }
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java]: Нужно подумать можем ли мы убрать 2 элса, а вниз просто оставить - response.sendRedirect("/"); на оба негативных случая(нет сессии/юзера)

**Review by AlexKievDeveloper — CHANGES_REQUESTED** [review]

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java [hunk 7]
```diff
@@ -35,6 +35,19 @@ public boolean add(String email, String password) {
...
     }
 
+    @Override
+    public boolean edit(String oldEmail, String newEmail, String newPassword){
+        boolean isEdit = false;
+        if (newEmail != null && newPassword != null && oldEmail != null){
+            User userToEdit = securityService
+                    .createUserWithHashedPassword(newEmail, newPassword.toCharArray());
+            isEdit = userDao.edit(oldEmail, userToEdit);
+            logger.debug("Attempt to update user with known email address in database with boolean result" +
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java]: Please add space after word result.

File: src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("text/html;charset=utf-8");
+        response.setStatus(HttpServletResponse.SC_OK);
+        logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_OK);
+        Map<String, String> attributes = new HashMap<>();
+        HttpSession session = request.getSession(false);
+        if (session != null){
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java]: Ctrl+Alt+L

File: src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                            logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
+                            attributes.put("message", "Edit is fail! Try again!");
+                        }
+                    } else {
+                        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                        logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
+                        attributes.put("message", "Sorry, old password isn't correct!");
+                    }
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java]: Please remove empty row

File: src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                }
+                attributes.put("email", email);
+                PageGenerator.getInstance().process("editProfile", attributes, response.getWriter());
+            } else {
+                response.sendRedirect("/signIn");
+            }
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java]: Please remove empty row

File: src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        } else {
+            response.sendRedirect("/signIn");
+        }
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java]: Please remove empty row

File: src/main/java/com/vinylteam/vinyl/web/servlets/HomeServlet.java [hunk 12]
```diff
@@ -18,13 +19,16 @@
...
-        Map<String, String> attributes = new HashMap<>();
-        User user = (User) request.getSession().getAttribute("user");
-        if (user != null) {
-            attributes.put("userRole", String.valueOf(user.getRole()));
-        }
         response.setStatus(HttpServletResponse.SC_OK);
         logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_OK);
+        Map<String, String> attributes = new HashMap<>();
+        HttpSession session = request.getSession(false);
+        if (session != null){
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/HomeServlet.java]: Ctrl+Alt+L

File: src/main/java/com/vinylteam/vinyl/web/servlets/ProfileServlet.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("text/html;charset=utf-8");
+        response.setStatus(HttpServletResponse.SC_OK);
+        logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_OK);
+        Map<String, String> attributes = new HashMap<>();
+        HttpSession session = request.getSession(false);
+        if (session != null){
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/ProfileServlet.java]: Ctrl+Alt+L

File: src/main/java/com/vinylteam/vinyl/web/servlets/SearchResultsServlet.java [hunk 14]
```diff
@@ -27,9 +28,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         String matcher = request.getParameter("matcher");
         List<Vinyl> filteredUniqueVinyls = vinylService.getManyFilteredUnique(matcher);
         attributes.put("searchWord", matcher);
-        User user = (User) request.getSession().getAttribute("user");
-        if (user != null) {
-            attributes.put("userRole", String.valueOf(user.getRole()));
+        HttpSession session = request.getSession(false);
+        if (session != null){
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SearchResultsServlet.java]: Ctrl+Alt+L

File: src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java [hunk 15]
```diff
@@ -30,9 +30,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         response.setContentType("text/html;charset=utf-8");
         response.setStatus(HttpServletResponse.SC_OK);
         Map<String, String> attributes = new HashMap<>();
-        User user = (User) request.getSession().getAttribute("user");
-        if (user != null) {
-            attributes.put("userRole", String.valueOf(user.getRole()));
+        HttpSession session = request.getSession(false);
+        if (session != null){
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/web/servlets/SignInServlet.java]: Ctrl+Alt+L

File: src/test/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDaoITest.java [hunk 16]
```diff
@@ -169,4 +169,37 @@ void addExistingWithNewPasswordTest() {
...
+        user.setSalt("");
+        user.setIterations(0);
+        user.setRole(Role.USER);
+        user.setStatus(true);
+
+        assertFalse(jdbcUserDao.edit(oldEmail, user));
+    }
+
+    @Test
+    @DisplayName("Edit non-existent user in db")
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDaoITest.java:189]: Maybe existing?

Comment (TarasOrlovskyi) [reply src/test/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDaoITest.java:189]: I think that 'non-existent' is also correct. Please, check it

**Review by AlexKievDeveloper — DISMISSED** [review]

File: src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+                } else {
+                    boolean checkOldPassword = securityService.checkPasswordAgainstUserPassword(user, oldPassword.toCharArray());
+                    if (checkOldPassword) {
+                        boolean isEdit = userService.edit(email, newEmail, newPassword);
+                        logger.debug("Got result of edit user by email with " +
+                                "passed email and password to db {'newEmail':{}, 'isEdit':{}}", newEmail, isEdit);
+                        if (isEdit) {
+                            response.setStatus(HttpServletResponse.SC_SEE_OTHER);
+                            logger.debug("Set response status to {'status':{}}", HttpServletResponse.SC_SEE_OTHER);
+                            attributes.put("message", "Your profile is edit successful.");
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/web/servlets/EditProfileServlet.java]: "Your profile is successfully changed."

File: src/test/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDaoITest.java (hunk 16 above)
Comment (Greengrer) [reply src/test/java/com/vinylteam/vinyl/dao/jdbc/JdbcUserDaoITest.java:189]: It is!

**Review by Greengrer — CHANGES_REQUESTED** [review]
It turns out I was wrong about reassigning mocks in beforeEach.
There exists reset(mock) that restarts mock without history of method calls and when(mock.function()).thenReturn() settings.

You could use:
private final HttpServletRequest mockedRequest = mock(HttpServletRequest.class);
private final InOrder inOrderRequest = inOrder(mockedRequest);

@BeforeEach
void beforeEach() {
reset(mockedRequest);
}

**Review by AlexKievDeveloper — APPROVED** [review]

**Review by Greengrer — APPROVED** [review]

**Pull request merged**